### DISCUSSION TITLE
Multiplication of multiple variables

### DISCUSSION THREAD

**Discussion by wendy444444** [yalmip/YALMIP#1104] [Optimization modelling]
II'm using yalmip and cplex for a milp project, but I'm having a problem. I know how to solve the problem of multiplying 0-1 variables by variables, but I want to solve the problem of "electricity price * electric power". Both of them are variables, and the result is WARNING: solver not applicable (CPLEX). I don't know what to do now. Please give me some suggestions. Thank you!

**Comment by johanlofberg** [comment]
If they are in the constraints, you simply cannot use that as cplex does not support bilinear or nonconvex quadratic constraints. Gurobi does (although you have to realize he problem becomes much much harder when extending to this problem class)

**Reply by wendy444444** [reply]
Thank you for your answer. I downloaded it as you said, but the program still can't run. As a result,"The continuous relaxation may be nonconvex. This means 
that the branching process is not guaranteed to find a
globally optimal solution, since the lower bound can be
invalid. Hence, do not trust the bound or the gap..."What should I do? Thank you very much for your help.

**Reply by johanlofberg** [reply]
You are now using the internal solver bnb, which as it states is not supposed to be used for nonconvex integer programs. You either failed to install gurobi, or you installed but the model you have has more than bilinear constraints making also gurobi unapplicable.

**Reply by wendy444444** [reply]
Thank you very much.Allow me to add that the results of using both solvers are“Solver not applicable ”.I may need to modify my constraints.

**Reply by johanlofberg** [reply]
Well,first you should understand why it isn't applicable, i.e. have you defined polynomial or semidefinite constraints by mistake etc. Assuming you are using a recent version of yalmip so it really connects to the nonconvex solvers in gurobi and cplex

**Reply by wendy444444** [reply]
I appreciate your help, but I'm sorry I didn't understand you. I tried to use mccormick relaxation to deal with bilinear problems, but it didn't work. The matlab version I use is 2018b, the yalmip version is before July 2018, the cplex version is IBM 12.10.0, and the gurobi version is 9.5.1. Does it matter?

**Comment by johanlofberg** [comment]
I asked you to display the constraints to see what constraint is outside the applicability of gurobi, and you start talking about McCormick which is a linear object (why are you reinventning the wheel...) 

and why on earth are you using a 4 year old version of YALMIP. I don't think that old version even supports the nonconvex gurobi module

**Reply by johanlofberg** [reply]
Unrecognized function or
variable 'kk'.
 
>>

**Reply by johanlofberg** [reply]
regardless, the model is obviously not bilinear

`Pij(i,:)=(PEbus(Ebranch(i,2),:).^2+QEbus(Ebranch(i,2),:).^2)./V(Ebranch(i,2),:).^2*Ebranch(i,4);`

You will have to reformulate that to a series of quadratic constraints to make gurobi applicable

**Reply by johanlofberg** [reply]
but the code is weird as you never use that expression anyway, so the problem is elsewhere

**Reply by johanlofberg** [reply]
with kk a double, gurobi solves this, so I don't know what you are asking about.

**Reply by johanlofberg** [reply]
result = 

  struct with fields:

    yalmipversion: '20210331'
    matlabversion: '9.9.0.1524771 (R2020b) Update 2'
       yalmiptime: 37.6148
       solvertime: 1.0652
             info: 'Successfully solved (GUROBI-NONCONVEX)'
          problem: 0

**Reply by wendy444444** [reply]
Maybe you have modified the code, because it still can't run successfully with me.

**Reply by johanlofberg** [reply]
the reason you cannot run it has already been answered above

**Reply by wendy444444** [reply]
Thank you. I got  it.
